src/rgw/rgw_d4n_datacache.cc

```diff
@@ -0,0 +1,478 @@
+#include "rgw_d4n_datacache.h"
+
+#define dout_subsys ceph_subsys_rgw
+#define dout_context g_ceph_context
+
+/* Base metadata and data fields should remain consistent */
+std::vector<std::string> baseFields{
+  "mtime",
+  "object_size",
+  "accounted_size",
+  "epoch",
+  "version_id",
+  "source_zone_short_id",
+  "bucket_count",
+  "bucket_size",
+  "user_quota.max_size",
+  "user_quota.max_objects",
+  "max_buckets",
+  "data"};
+
+std::vector< std::pair<std::string, std::string> > RGWD4NCache::buildObject(rgw::sal::Attrs* binary) {
+  std::vector< std::pair<std::string, std::string> > values;
+  rgw::sal::Attrs::iterator attrs;
+ 
+  /* Convert to vector */
+  if (binary != NULL) {
+    for (attrs = binary->begin(); attrs != binary->end(); ++attrs) {
+      values.push_back(std::make_pair(attrs->first, attrs->second.to_str()));
+    }
+  } 
+
+  return values; 
+}
+
+int RGWD4NCache::findClient(cpp_redis::client *client) { 
+  if (client->is_connected())
+    return 0;
+
+  if (host == "" || port == 0) { 
+    dout(10) << "RGW D4N Cache: D4N cache endpoint was not configured correctly" << dendl;
+    return EDESTADDRREQ;
+  }
+
+  client->connect(host, port, nullptr);
+
+  if (!client->is_connected())
+    return ECONNREFUSED;
+
+  return 0;
+}
+
+int RGWD4NCache::existKey(std::string key) { 
+  int result = -1;
+  std::vector<std::string> keys;
+  keys.push_back(key);
+
+  if (!client.is_connected()) {
+    return result;
+  }
+
+  try {
+    client.exists(keys, [&result](cpp_redis::reply &reply) {
+      if (reply.is_integer()) {
+        result = reply.as_integer();
+      }
+    });
+
+    client.sync_commit(std::chrono::milliseconds(1000));
+  } catch(std::exception &e) {}
```

Comment:
@cbodley Would an exception here shut down the RGW, if so is that the behaviour we want? Don't we just want the value of `result` to be returned?